lib/shared/files.js

```diff
@@ -0,0 +1,167 @@
+ * @returns {Object} file metadata
+ *
+ * @example
+ * const file = getFileMetadata('/home/user')
+ * console.log(`Is ${file.basename} a directory? ${file.isDirectory}`)
+ */
+exports.getFileMetadataSync = (dirname, basename = '') => {
+  // TODO(Shou): use path.parse object information here
+  const fullpath = path.join(dirname, basename)
+  const pathObj = path.parse(fullpath)
+
+  // TODO(Shou): this is not true for Windows, figure out Windows hidden files
+  const isHidden = pathObj.base.startsWith('.')
+  const stats = fs.lstatSync(fullpath)
+
+  return {
+    basename: pathObj.base,
+    dirname: pathObj.dir,
+    fullpath,
+    extension: pathObj.ext.replace('.', ''),
+    name: pathObj.name,
+    isDirectory: stats.isDirectory(),
+    isHidden,
+    size: stats.size
+  }
+}
+
+exports.getFileMetadataAsync = (dirname, basename = '') => {
+  // TODO(Shou): use path.parse object information here
+  const fullpath = path.join(dirname, basename)
+  const pathObj = path.parse(fullpath)
+
+  // NOTE(Shou): this is not true for Windows
+  const isHidden = pathObj.base.startsWith('.')
+  return fs.lstatAsync(fullpath).then((stats) => {
+    return {
+      basename: pathObj.base,
+      dirname: pathObj.dir,
+      fullpath,
+      extension: pathObj.ext.replace('.', ''),
+      name: pathObj.name,
+      isDirectory: stats.isDirectory(),
+      isHidden,
+      size: stats.size
+    }
+  })
+}
+
+exports.getAllFilesMetadataAsync = (dirname, basenames) => {
+  return Bluebird.map(basenames, (basename) => {
+    return exports.getFileMetadataAsync(dirname, basename)
+  })
+}
+
+/**
+ * @summary Get directory contents of a path
+ * @function
+ * @private
+ *
+ * @param {String} dirname - directory path
+ * @returns {Promise<Array<Object>>}
+ *
+ * @example
+ * const files = getDirectoryContents('/home/user')
+ * files.map((file) => {
+ *   if (file.isHidden) {
+ *     return `Hidden: ${file.fullpath}`
+ *   }
+ *
+ *   if (file.isDirectory) {
+ *     return `Directory: ${file.fullpath}`
+ *   }
+ *
+ *   return `File: ${file.basename} in ${dirname}`
+ * })
+ */
+exports.getDirectoryContents = (dirname) => {
+  return fs.readdirAsync(dirname).then((contents) => {
+    return Bluebird.map(contents, (basename) => {
+      return exports.getFileMetadata(dirname, basename)
+    })
+  })
+}
+
+exports.getDirectory = (dirname) => {
+  return fs.readdirAsync(dirname)
+}
+
+/**
+ * @summary Get all subpaths contained in a path
+ * @function
+ * @private
+ *
+ * @param {String} fullpath - path string
+ * @returns {Array<Object>} - all subpaths as file objects
+ *
+ * @example
+ * const subpaths = files.subpaths('/home/user/Downloads')
+ * console.log(subpaths.map(file => file.fullpath))
+ * > [ '/', '/home', '/home/user', '/home/user/Downloads' ]
+ */
+exports.subpaths = (fullpath) => {
+  if (!_.isString(fullpath)) {
+    return null
+  }
+
+  const dirs = _.compact(_.update(fullpath.split(path.sep), [ '0' ], (root) => {
```

Comment:
This method should be checking with `path.isAbsolute()`, and split with `path.basename()`and `path.dirname()` – as splitting on `path.sep` can break in horrible ways with escaped chars, and on Windows, both `/` and `\` are legit path separators

Reply:
Couldn't we just `path.resolve` instead of checking with `path.isAbsolute`?

Reply:
Sure, if you always want an absolute path